### PR TITLE
Search within aor

### PR DESCRIPTION
**Pull request by jabrah**
Some changes I made to get the search within widget to work with my data. Most of the changes around getting he canvas ID and annotation (hit) label. These changes should apply to the cases where the search service returns a hits object and when it does not. They should work for your data as well, however, there might be some differences in how we want things labelled.

Let me know if this breaks anything for you.

### CODE REVIEW

**Comment by jeffreycwitt** [comment]
Hey John, thanks for this. 

Looks like a lot of good improvements. I've pulled down your changes to a local branch and i'm working with them now.

I'm a little worried about lines 65, 101, and 106, namely the call to `resource.label`.

Here I was trying to get the label of the canvas, on which the annotation is being painted. 
Lines 65 and 106 are looking for a "label" property on a given annotation or hit, but currently I'm not aware of a label property for the actual `annotation` or the `hit` object. So this call for label is coming up empty with my search results. (see for example: http://exist.scta.info/exist/apps/scta/iiif/pp-sorb/search?q=fides)

It really depends on what we want. Do we want to find a label for an individual search result or for the canvas on which the search result falls?

I was trying to do the latter, namely find a label for the canvas on which the search result falls (that's why I was ripping it out of the canvas id). 

The better thing to do is access the hash of canvasid to canvas object properties that Mirador builds after loading a manifest. Drew said there is away to access this. But i'm not sure. Maybe I can ask him.

In other words, for line 65 at least, I think I want to do something like:

canvaslabel = this.manifest.getLabel(canvasid).
or
canvaslabel = this.manifestHash.getLabel(canvasid).

Let me know what you think.

**Comment by jabrah** [comment]
That is a good question, what do we want to display in the search hit label? My understanding is that the search spec has no say regarding UI labels, leaving this detail up to the implementation.

The reason behind my having individual labels per hit (by grabbing the associated annotation label) is for the purposes of making it easier to identify the hit on the page. My data does not have spatial coordinates, so all annotations are target the whole canvas. On top of that, there are different categories of annotations that the scholars on my project care about. For example, they want to be able to differentiate between transcription of underlined text and transcription of a marginal note made by a reader. For my use case of search, it does not seem like having only the canvas as the label would give a user enough information to effectively find the search result on the page.

However, going back to the Presentation spec and I can see your point. There is a nice table at the end of the spec that gives all the [**Descriptive rights and properties**](http://iiif.io/api/presentation/2.1/#b-summary-of-metadata-requirements). A few objects have a label required, but annotation is among the objects that has label as optional. So the annotation can have a label, but does not have to. 

Would you be willing to have the search hit label such that it will be set to the annotation label if present. If not, set it to the canvas label?

Regarding the retrieval of the canvas label, I do recall that conversation between you and Drew, though I do not remember the exact contents. I can look into this tomorrow.

**Comment by jabrah** [comment]
So I do not know where Drew's helper method is that he talked about in Gitter regarding the canvas hash stuff. However, I think he is just using an associative array in Javascript. Here is a code example of what I was thinking:

``` javascript
$.Manifest.prototype = {
    init: function(manifestUri) {
      var _this = this;
      this.request = jQuery.ajax({
        url: manifestUri,
        dataType: 'json',
        async: true
      });

      this.request.done(function(jsonLd) {
        _this.jsonLd = jsonLd;
        _this.buildCanvasMap();
      });
    },
    buildCanvasMap: function() {
      var _this = this;
      this.canvasMap = {};

      this.getCanvases().forEach(function(canvas) {
        _this.canvasMap[canvas['@id']] = canvas;
      });
    },

    // ...

    getCanvasLabel: function(canvasId) {
      canvasId = canvasId.split('#')[0];  // Remove region fragment if necessary
      return this.canvasMap[canvasId].label;
    }
  }
```

Once the manifest is loaded, an object is created similar to a Dictionary or a Map in other languages, where the whole canvas object is indexed by its ID. In this way, you can get easy access to a canvas if you have its ID by just calling: `canvasMap[canvasId]` There is a trade off with this way of accessing the canvases, however. Canvas access is faster this way, but it will use some extra memory and the initial load of the manifest will be slightly longer.

**Comment by jeffreycwitt** [comment]
Yes, I like that. I think Drew might have simply said we need to make a helper in the way you have proposed

I also like your idea of FIRST looking for a label on the annotation, and THEN if no label is found falling back to look for the label on the canvas being annotated.

**Comment by jabrah** [comment]
OK, I just found out that Github will automatically add commits to a pull request :)

I added some changes to how the labels are found, defaulting to the annotation label. If the annotation does not have a label, then the canvas label will be displayed. In the manifest object, all canvases are first put into an array, retrievable by its canvas ID, making canvas look up easier.

**Comment by jeffreycwitt** [comment]
Hi John, sounds, good. I tried to take a look at this afternoon. But my search results are no longer coming up. I wasn't able to figure out the problem entirely, but I think i was getting a hiccup at the attempt to get the canvas label. Try using this manifest: http://scta.info/iiif/pp-svict/manifest (I was searching for the word "fides" which should give you these search results: exist.scta.info/exist/apps/scta/iiif/pp-sorb/search?q=fides)

**Comment by jabrah** [comment]
Sorry, things have been a bit interesting this week. I will troubleshoot against your search service when I get time, probably tomorrow.

**Comment by jabrah** [comment]
I ran new code against your search service, searching for 'fides'. It looks like a couple of the results point to a canvas: 'http://scta.info/iiif/pp-svict/canvas/SV' I could be mistaken, but I do not think this is intended, as there does not seem to be a canvas in this particular manifest with that ID. Since no canvas was found in the manifest, current behavior is that Mirador will default back to the first page in the sequence.

Here is an example result from your search service, please forgive encoding errors from copying :). 

``` javascript
        {
            "resource": {
                "@type": "cnt:ContentAsText",
                "chars": "<p>Ex quo \n            sequitur non esse   necessariam  necessariam   Reducere illud ad aliquem \n            modum naturaliter nobis notum quia  <span style=\"background-color: yellow;\">fides </span> non obligat \n            nos nisi ad credendum quod corpus christi sit ibi in hostia \n            et in qualibet eius parte,   sed hoc est supernaturaliter   ideo \n            modus iste existendi non debet Reduci ad modum naturalem \n            nobis cognitum quod si Reduceretur hoc esset magis \n            ampliare credulitatem quam fidem[?] nos obligat et per consequens \n            diceretur et per consequens doctores istam \n            materiam particularitates transgrediuntur methas[?] artis \n            obligatoriae fidei quia propter hoc concedunt maiora \n            inconvenientissima quam difficultas principalis,   et ideo plures nega  verunt \n          quod unum corpus possit esse dime??ve[?] in pluri  bus \n          locis ut  <span style=\"font-variant: small-caps\">sanctus thomas </span>,   quia ex fide non habe  mus \n          nisi quod idem corpus christi potest esse in pluribus locis et \n            Realiter existit.   Ex quo non sequitur quod idem corpus \n            possit esse dimen??ve[?] in pluribus locis eo quod consequens est \n            maioris difficultatis quam antecedens. </p>"
            },
            "motivation": "sc:painting",
            "@type": "oa:Annotation",
            "@id": "http://scta.info/iiif/pp-svict/search/annotations/prIV-eqsdqa",
            "on": "http://scta.info/iiif/pp-svict/canvas/SV"
        }
```

This does bring up a good point, though, of letting this fail more gracefully. In this case where no canvas was found for the given ID. I think these search results should still be shown, however, the labels will be more tricky to determine, if it is needed.

I can see some alternative labels to display if the canvas is found not exist.
- We can simply ignore these results. If no canvas was found in the manifest, can we assume that the canvas does not exist (for purposes of SearchWithin)? If this is done, these results will just not be displayed.
- We can have a blank label. This, I think is kind of useless as there is no real context for the result and Mirador will not be able to display the image of a canvas that it cannot find.
- Display the bare URI ID as the label. Not great for the user, as URIs are not designed to be human readable.

I am personally in favor of the first option. I do not think that we should try to extract the canvas name from the URI ID, since the ID structure defined in the IIIF Presentation spec is only recommended, not mandatory.

**Comment by jeffreycwitt** [comment]
@jabrah sorry for the delayed response. I’ve been swamped with the start of school. 

Anyway, I really like your proposal. The reason my search results are returning a canvas that doesn’t exist is because the data is incomplete. 

But all the more reason to need a graceful fallback as you propose. 

If you wanna go ahead and push the new commit I’ll pull it down on my machine and see how it works.

**Comment by jabrah** [comment]
I pushed some code that is now included in this pull request. Hopefully this will make it more robust, so that it will display results even if there are errors in parsing, such as canvases that do not exist in the manifest. I tried the new code on my own data as well as your example you gave above, seemed to be OK.

On an unrelated note, I have started trying to pull out the Search Within code from _window.js_. Once this PR is settled, I will see about getting that into shape for a separate pull request.

**Comment by jeffreycwitt** [comment]
John this is awesome. It seems to be working really well.

**Comment by jeffreycwitt** [comment]
I notice that my search results are not longer displaying for results that don't have an identified "canvas". Is that correct? This probably is the desired behavior though, because search results without a valid "canvasid" really should be considered valid results. 

So I like the way it works now, where it only displays those results that have a valid id.

**Comment by jeffreycwitt** [comment]
Do you think I should merge it in now?

**Comment by jeffreycwitt** [comment]
Live working version here [http://miradorlab.scta.info/#44c1a2c1-2536-4031-873c-f80cfc6db85d](http://miradorlab.scta.info/#44c1a2c1-2536-4031-873c-f80cfc6db85d)

Note also that after I merged the pull request, I added one more commit removing the "close" button in the template. A user can open a close simply by toggling the search icon.

**Comment by jabrah** [comment]
Excellent, the demo looks good!

Yeah, my idea with the fallback behavior was to simply omit any search results that were not attached to a known canvas. Without the link to a canvas, I would think that navigation to the result location is likely not possible in a general sense. Not being a researcher, I can't judge how useful these results would be, since they would lack the context of a page in the manifest. If they are valuable to have in the results list, it should fairly easy to add them back in.
